### DISCUSSION TITLE
Font change on beamer slides when building with pandoc 3.5

### DISCUSSION THREAD

**Discussion by pck1980** [jgm/pandoc#10297] [Q&A]
I just realized my freshly built beamer slides look different than they used to. The only thing I changed (as far as I remember) was updating pandoc from 3.4 to 3.5
I use the Metropolis theme which used to look like in the first screenshot, but now it looks as shown in the second screenshot.
<img width="272" alt="old" src="https://github.com/user-attachments/assets/8ddf107f-2806-439d-a839-c01235d6dbfc">
<img width="331" alt="new" src="https://github.com/user-attachments/assets/605e6047-6d5b-4f4b-836f-f47d36e4bd60">

The font change does not only affect the title, but all slides.

My markdown header looks like this:

```
---
title: blah
author: blah
date: blah
institute: blah 
theme: metropolis
keywords: [blah]
---
```

Forgot to mention: I built the slides with the command `pandoc -s -t beamer --pdf-engine=lualatex -o slides.pdf slides.md -V theme:metropolis -V themeoptions:subsectionpage=progressbar -V lang=de`

The changelog mentions several changes to Beamer output, but I have no idea what I have to do to get back to the old appearance.
Thanks in advance for any help.

**Comment by jgm** [comment]
Which of these matches the appearance metropolis is supposed to have?

**Reply by jgm** [reply]
I think that to diagnose this, you can do:

- use pandoc 3.5: `pandoc test.md -t beamer -s -o test-3.5.tex -V theme:metropolis`
- use pandoc 3.4: `pandoc test.md -t beamer -s -o test-3.4.tex -V theme:metropolis`
- `diff -u test-3.4.tex test-3.5.tex`

This should give us a full accounting of what has changed, and we can then test changes to test-3.5.tex that would restore the old behavior (assuming it is the proper one).

**Reply by pck1980** [reply]
I think the „old“ look is the correct one. 

Can I simply install 3.4 again on macOS to go back to that version?

**Reply by pck1980** [reply]
I tried building a test document with both pandoc 3.4 and 3.5 as you suggested. Here's the output of diff:

[diff.txt](https://github.com/user-attachments/files/17421157/diff.txt)

And the test file I used:

[test.md](https://github.com/user-attachments/files/17421204/test.md)

Going back to 3.4 also confirmed that it was indeed the update to 3.5 that changed the appearance of my slides. With 3.4, I get the old (correct) look again.

**Reply by pck1980** [reply]
I just realized that both tex files result in a "wrong looking" pdf when processing them with pdflatex, whereas lualatex creates a correct-looking pdf from test-3.4.tex, but a wrong looking from test-3.5.tex

**Comment by pck1980** [comment]
It seems that moving the line `\usetheme[]{metropolis}` down to where it was with version 3.4 solves the issue.

**Reply by jgm** [reply]
good to know.  I can look into fixing that.

**Reply by twsh** [reply]
One thing that you could do without changing the template is to choose Fira Sans using 'fontfamily', perhaps like this: `--variable=fontfamily:FiraSans --variable=fontfamilyoptions:semibold --variable=fontfamilyoptions:light`. That should work with pdflatex. (I took those options from the documentation for [Moloch](https://github.com/jolars/moloch), a fork of Metropolis.)

Otherwise, it does seem that the problem is that the new template loads lmodern.sty after the template is set which means that lmodern.sty overrides the font setting done by Metropolis.

**Reply by twsh** [reply]
I noticed that the fonts.latex template has this comment: '$-- Set default font before Beamer theme so the theme can override it'. The default.beamer template has the comment '$-- Set Beamer theme before user font settings so they can override theme'. I'm not sure what the right thing to do is. I can imagine someone who wants to use Metropolis with a different font who would want the font to be set after the theme is. On the other hand, Metropolis won't work as expected now.

**Reply by jgm** [reply]
In an earlier iteration of the new template, I had two separate partials -- fonts and font-settings -- so that the beamer theme could go between them.  I guess I convinced myself by experimentation that this wasn't necessary, but I may have been missing something.  We could go back to that.

**Reply by twsh** [reply]
Moving the themes seems like the right thing to me. Couldn't it just be moved to after common.latex() instead of splitting fonts.latex()? I think that would fix Metropolis.

If you like, I can try to do a PR next week after testing a few theme and font combinations.

**Reply by jgm** [reply]
IF I recall, the reason we split before was to allow people to use a theme like metropolis with a different font than metropolis specifies.  Would your solution allow that?

**Reply by twsh** [reply]
Only if they use pdflatex as their pdf-engine. So that would be a reason to split the fonts partial and put the theme between them.

**Reply by twsh** [reply]
I can see where you intended to split fonts.latex. I tested a few different combinations with a split there. It fixes this problem and supports overriding Metropolis fonts.

One question I have now is what to call the two parts of fonts.latex. I have currently called them 'beforethemefonts.latex' and 'afterthemefonts.latex'.

Another is how to handle this in default.latex. The idea I have tried is to change fonts.latex so that it calls the partials 'beforethemefonts.latex' and 'afterthemefonts.latex'. I don't know if that is overcomplicating things.

**Reply by jgm** [reply]
I did some testing and I'm seeing this:

Putting `\usetheme[]{metropolis}` right after `\usepackage{lmodern}` yields the result you want *only* if lualatex is used. With xelatex or pdflatex, we still get the wrong font.

**Reply by twsh** [reply]
There shouldn't be a change with pdflatex. Metropolis tries to change the font when xelatex or lualatex are used.

When I try, I do see the correct change to Fira Sans with both xelatex and lualatex.

My commands:

* `pandoc --pdf-engine=lualatex --template=default.beamer --to=beamer --variable=theme:metropolis --output=metropolis.pdf`
* `pandoc --pdf-engine=xelatex --template=default.beamer --to=beamer --variable=theme:metropolis --output=metropolis.pdf`

Here is the LaTeX I am getting: https://gist.github.com/twsh/ce933cb20938645b1c3eed3776c06fed

**Reply by jgm** [reply]
Weird. I downloaded your latex and compiled it with xelatex and lualatex. With lualatex I got Fira Sans but with xelatex I got a different font (presumably lmodern sans)?

**Reply by jgm** [reply]
In the xelatex run I see an error message
```
Package beamerthememetropolis Warning: Could not find Fira Sans fonts on input 
line 95.
```
I wonder why this would work with your xelatex setup but not mine?  Note: I don't have "Fira Sans" installed on my system, but this doesn't seem to prevent lualatex from finding it (perhaps in the metropolis style files)?

**Reply by twsh** [reply]
I can confirm that trying it on a system without Fira Sans deliberately added, I see what you saw. I don't know how lualatex manages to find it. I do have a full texlive installation (on both systems) and that includes the Fira fonts.

**Reply by jgm** [reply]
OK, @badumont has clarified this.

So, I guess it looks like we should re-split the `fonts` partial.  Originally instead of `fonts` I had two partials, `fonts` and `font-settings`.  We could go back to that, but for backwards compatibility it might be wiser to create two new partials and include them from `fonts`. Then custom templates based on pandoc 3.5 would not need to be modified.  The cost, though, would be an uglier and more indirect system.  @tarleb do you have thougths on this?

**Reply by tarleb** [reply]
I can't really comment on possible alternatives, but `fonts` and `font-settings` seems  clean enough IMHO. My only thought was that a name like `fonts-override` could help to clarify the intend for that partial, but I'm not sure.

**Reply by jgm** [reply]
My question wasn't really about the names, but about whether we should adopt a suboptimal plan (making the `fonts` partial include two new partials, as opposed to using the two partials directly in default.latex) in order to retain backwards compatibility with custom templates based on pandoc 3.5.

**Reply by tarleb** [reply]
Ah. I'd rather break backwards compatibility then. Projects like Quarto expect a certain instability and maturation period for new features, so I'd say it's fine to change it now.

**Comment by badumont** [comment]
> I wonder why this would work with your xelatex setup but not mine?  Note: I don't have "Fira Sans" installed on my system, but this doesn't seem to prevent lualatex from finding it (perhaps in the metropolis style files)?

LuaLaTeX probably finds it in your TeXLive directory (or whatever TeX distribution you use). But, as the fontspec manual says: “In XeTeX, fonts fond in the TEXMF tree can be loaded in Windows and Linux, but not on MacOS.”

**Comment by jgm** [comment]
I have pushed a fix that splits the fonts partial into two parts.
